pkg/runner/runner.go

```diff
@@ -0,0 +1,346 @@
+}
+
+type tracerInterface interface {
+	InitWithGadgetInfo(*types.GadgetInfo) error
+	RunWithoutGadgetContext(logger logger.Logger, ctx context.Context, timeout time.Duration, params *params.Params) error
+	Close()
+}
+
+// NewRunner creates a new runner with the given image and options
+func NewRunner(image string, opts ...func(*RunnerOpts)) (*Runner, error) {
+	return newRunner(image, opts...)
+}
+
+// NewRunnerFromBytes creates a new runner with the given eBPF program bytes and options
+// The Metadata will be inferred from the eBPF program
+func NewRunnerFromBytes(ebpfBytes []byte, opts ...func(*RunnerOpts)) (*Runner, error) {
+	return newRunner(ebpfBytes, opts...)
+}
+
+func newRunner(program any, opts ...func(*RunnerOpts)) (*Runner, error) {
+	gadgetDesc := tracer.GadgetDesc{}
+	runGadget, err := gadgetDesc.NewInstance()
+	if err != nil {
+		return nil, fmt.Errorf("creating run gadgetDesc instance: %w", err)
+	}
+	tracerInstance := runGadget.(*tracer.Tracer)
+
+	r := newRunnerFromTracer(tracerInstance, opts...)
+
+	gadgetInfo, err := getGadgetInfo(program, &gadgetDesc, &r.RunnerOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	err = r.tracer.InitWithGadgetInfo(gadgetInfo)
+	if err != nil {
+		return nil, fmt.Errorf("initializing run gadget: %w", err)
+	}
+
+	// TODO: Add functions to set the params
+	r.RunnerOpts.ebpfParams = gadgetInfo.GadgetMetadata.EBPFParams
+
+	// Set Event Handler
+	conv := gadgetDesc.JSONConverter(gadgetInfo, &r.eventBuffer)
+	tracerInstance.SetEventHandler(func(ev *types.Event) {
+		conv(ev)
+	})
+	tracerInstance.SetEventHandlerArray(func(ev []*types.Event) {
+		conv(ev)
+	})
+
+	return r, nil
+}
+
+func getGadgetInfo(program any, gadgetDesc *tracer.GadgetDesc, RunnerOpts *RunnerOpts) (*types.GadgetInfo, error) {
+	switch p := program.(type) {
+	case string:
+		gadgetInfo, err := gadgetDesc.GetGadgetInfoWithRunOpts(p, RunnerOpts.RunOpts, logger.DefaultLogger())
+		if err != nil {
+			return nil, fmt.Errorf("getting gadget info: %w", err)
+		}
+		return gadgetInfo, nil
+	case []byte:
+		gadgetInfo, err := gadgetDesc.GetGadgetInfoFromBytesRunOpts(p, RunnerOpts.RunOpts, logger.DefaultLogger())
+		if err != nil {
+			return nil, fmt.Errorf("getting gadget info: %w", err)
+		}
+		return gadgetInfo, nil
+	default:
+		return nil, fmt.Errorf("invalid program type")
+	}
+}
+
+func newRunnerFromTracer(t tracerInterface, opts ...func(*RunnerOpts)) *Runner {
+	r := &Runner{
+		tracer: t,
+		RunnerOpts: RunnerOpts{
+			RunOpts: tracer.RunOpts{
+				ValidateMetadata: true,
+				PullPolicy:       oci.PullImageMissing,
+				AuthOpts: &oci.AuthOptions{
+					AuthFile: oci.DefaultAuthFile,
+				},
+			},
+			Timeout: 0,
+			context: context.Background(),
+		},
+		state:      RunnerStateNotStarted,
+		stateMutex: sync.Mutex{},
+
+		eventBuffer: eventBuffer{
+			buffer: []string{},
+			// This is definitely an invalid JSON object and should never occur as a normal event
+			eventPoisonPill: "][}]]][[[][}[]]]]]]{[[][[[[}}{}{{[]]]{",
```

Comment:
Why not just using an empty string to detect the end of the stream?

Reply:
Right. The run tracer with the json conversion should never give us an empty string